### PR TITLE
Do not count attribute docstrings for WPS226

### LINKED ISSUE #3805: `WPS226` counts PEP 258 attribute docstrings
With docstrings excluded from `WPS226` (#3804), functions, classes and modules are fine, but the attribute form still counts:

```python
class Some:
    first = 1
    """Some field docs."""

    second = 2
    """Some field docs."""

    third = 3
    """Some field docs."""

    fourth = 4
    """Some field docs."""
```

```
attr_doc.py:3:5: WPS226 Found string literal over-use: 'Some field docs.' 4 > 3
```

Same reasoning as #3803: extracting that string into a constant stops it from being documentation. PEP 258 calls it an attribute docstring and Sphinx `autodoc` picks it up.

It isn't covered by the fix for #3803 because that one defines a docstring as `body[0]` of the nearest module, class or function, which is what `is_doc_string()` in `logic/tree/strings.py` has always meant. An attribute docstring is the statement after an assignment, so it never matches.

I'm genuinely unsure this is worth fixing. This project documents its own module attributes with `#:` comments (147 of those lines in `wemake_python_styleguide/`), so the style guide doesn't obviously owe support to the other convention. But it's the same false positive as #3803, just one line lower.

If you do want it, the change looks narrow: also skip a string `ast.Expr` that directly follows an assignment in the same body. I can pick it up.

### PR DESCRIPTION
**Pull request by vyhuholl**
# I have made things!

`WPS226` still counted PEP 258 attribute docstrings after #3804. Four class attributes sharing one:

```
attr_doc.py:3:5: WPS226 Found string literal over-use: 'Some field docs.' 4 > 3
```

Same reasoning as #3803. The string documents the assignment above it, so moving it into a named constant would stop it from being documentation.

`is_doc_string_value()` accepted exactly one position, `body[0]` of the nearest module, class or function. It now accepts a second, and the lookup moved into a helper:

```python
def _is_doc_string_place(body, statement) -> bool:
    """Docstrings open a definition's body or follow an assignment."""
    if body[0] is statement:
        return True
    return any(
        current is statement and isinstance(previous, AssignNodes)
        for previous, current in itertools.pairwise(body)
    )
```

`AssignNodes` is the existing `(ast.Assign, ast.AnnAssign)` tuple, so `first = 1`, `second: int = 2` and a bare `third: int` all work. I left `ast.AugAssign` out, since `count += 1` followed by a string isn't documenting anything.

The search stays inside the context's own body. `self.third = 3` in `__init__` is covered, because the context there is the method. A string after an assignment nested inside an `if` still counts, same as before, and so does one that follows anything other than an assignment. Both have tests.

One thing that looks odd in the diff: `is_doc_string_value()` now ends with `context is not None and _is_doc_string_place(...)` instead of an early return. Writing it as `if context is None: return False` drops coverage below 100%, because `get_context()` returns `None` only for the module node itself, and that can never be a docstring statement.

## AI Policy

- [x] I have read and agree to the [AI Policy](https://github.com/wemake-services/wemake-python-styleguide/blob/master/.github/AI_POLICY.md), removed any "Co-Authored-By" lines attributing coding agents, and manually reviewed the final result

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #3805 

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

**Review by sobolevn — COMMENTED** [review]
Please, also test type-aliases, see https://discuss.python.org/t/docstrings-for-type-aliases/108901 for the context

**Comment by vyhuholl** [comment]
All addressed, and two of the notes turned into bigger changes than they looked.

`views.py` is now a package: `base.py`, `body.py`, `cookie.py`, with
`__init__.py` re-exporting everything, so `dmr.security.jwt.views.X` keeps
working. The shared bases moved to `base.py` and lost their leading underscore
(`BaseTokenController`, `BaseRefreshTokenController`), because pyright rejects
importing `_`-prefixed names across modules. They are not re-exported from the
package, so the public surface is unchanged.

`check_auth` now takes `(user: AbstractBaseUser, token: JWToken)` on the cookie
controllers and on the four body ones as well. Two more things were needed
before the mixins actually worked:

- `RefreshTokenSyncController` and `RefreshTokenAsyncController` had no
  `get_user`, the lookup was inlined into `refresh`, and `blocklist()` calls it.
  It is now a method there too, like the verify controllers have.
- `_BaseBlocklistMixin.__init__` assumed an auth object and set
  `self.require_claims`, which a controller does not have. It now skips that for
  controllers. A token without `jti` is still rejected, `token_jti` raises for it.

There are five tests for this in `test_blocklist_views.py`: a blocklisted token
cannot refresh, sync and async, cookie and body flow, and a token without `jti`
is rejected. Documented in the blocklist section with an example.

On `secure`: I would keep `True`. Browsers treat `http://localhost` and
`http://127.0.0.1` as secure contexts, so `Secure` cookies are stored there and
local development works as is. `not settings.DEBUG` would make the security of
the flow depend on a flag that is wrong in staging often enough. The case it
does break is a dev environment served over plain HTTP on some other host name,
and the docs now say to set `jwt_cookie_secure = settings.DEBUG is False`
yourself there.

The rest:

- `reverse_lazy` in the test app. `CookieSpec.path` and `NewCookie.path` now
  accept lazy strings, otherwise users have to hardcode the refresh url the same
  way. The docs examples use `reverse_lazy` too.
- `cast` replaced with `type: ignore` in both `make_api_response`.
- `@sensitive_variables` dropped from the sync `login` and `refresh`, added to
  the async `get_user` and to both logout endpoints.
- The inline code block in `jwt.rst` is a separate example file now.
- Test for a redefined `validate_spec`. Writing it showed that an override had
  no way to keep the `Cache-Control` header without reaching for a private
  constant, so the cookie controllers got `response_headers` and
  `response_headers_spec` next to the cookie spec hooks.

Breaking changes and the new methods are in the changelog.

(CI is failing because it's failing on master now, cause — commit 900b7d77)

**Comment by sobolevn** [comment]
🤔

**Comment by vyhuholl** [comment]
Sorry, that was comment for the wrong PR. 🙈

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py [hunk 1]
```diff
@@ -169,6 +169,50 @@ def fourth():
...
+def second():
+    print(2)
+    {0}
+
+def third():
+    print(3)
+    {0}
+
+def fourth():
+    print(4)
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py]:
```suggestion
    x = 1
```

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py [hunk 2]
```diff
@@ -169,6 +169,50 @@ def fourth():
...
+def first():
+    print(1)
+    {0}
+
+def second():
+    print(2)
+    {0}
+
+def third():
+    print(3)
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py]:
```suggestion
    y: int = 0
```

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py [hunk 3]
```diff
@@ -169,6 +169,50 @@ def fourth():
...
+{0}
+
+second: int = 2
+{0}
+
+third: int
+{0}
+"""
+
+class_attribute_docstrings = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py:185]: please, don't forget about `type` aliases on `3.12+`, they should also be treated as targets for docstrings

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py [hunk 4]
```diff
@@ -169,6 +169,50 @@ def fourth():
...
+        {0}
+"""
+
+not_an_attribute_docstring = """
+def first():
+    print(1)
+    {0}
+
+def second():
+    print(2)
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py]:
```suggestion
    x, y = call()
```

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py [hunk 5]
```diff
@@ -169,6 +169,50 @@ def fourth():
...
+    {0}
+
+    def __init__(self):
+        self.third = 3
+        {0}
+"""
+
+not_an_attribute_docstring = """
+def first():
+    print(1)
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py]:
```suggestion
    x = y = 1
```

File: tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py (hunk 3 above)
Comment (vyhuholl) [reply tests/test_visitors/test_ast/test_complexity/test_overuses/test_overused_string.py:185]: Done in [#957cbea3 ](https://github.com/vyhuholl/wemake-python-styleguide/commit/957cbea3cc867b9bc59e0b727587ebd3fdf5b952)

File: wemake_python_styleguide/logic/tree/strings.py [hunk 6]
```diff
@@ -26,12 +33,52 @@ def is_doc_string_value(node: ast.AST) -> bool:
...
+        return isinstance(statement, nodes.TypeAlias)
+    if len(targets) != 1:  # `x = y = 1` defines no single attribute
+        return False
+    if isinstance(context, FunctionNodes):
+        # Locals are not attributes, only `self.some = 1` is one.
+        return isinstance(targets[0], ast.Attribute)
+    return isinstance(targets[0], _AttributeTargets)
+
+
+def _get_assign_targets(statement: ast.stmt) -> list[ast.expr]:
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/tree/strings.py]: Will `get_assign_targets` work here?

Comment (vyhuholl) [reply wemake_python_styleguide/logic/tree/strings.py]: Yes, it works — committed as https://github.com/vyhuholl/wemake-python-styleguide/commit/0e593d0cc76b39532ddce08b1314e6255a6469d4. The private helper was redundant. Thank you for noticing!

File: wemake_python_styleguide/logic/tree/strings.py [hunk 7]
```diff
@@ -26,12 +34,43 @@ def is_doc_string_value(node: ast.AST) -> bool:
...
+def _is_documented(statement: ast.stmt, context: ContextNodes) -> bool:
+    """Only type aliases and single attribute definitions are documented."""
+    if not isinstance(statement, AssignNodes):
+        return isinstance(statement, nodes.TypeAlias)  # `type X = int`
+    targets = get_assign_targets(statement)
+    if len(targets) != 1:  # `x = y = 1` defines no single attribute
+        return False
+    if isinstance(context, FunctionNodes):
+        # Locals are not attributes, only `self.some = 1` is one.
+        return isinstance(targets[0], ast.Attribute)
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/tree/strings.py]: this does not seem fully correct. currently this code:

```python
def some():
    x.y = 1
    """Docstring."""
```

counts as a docstring, while it is clearly not 🤔 
not sure what is the best way forward here. 

maybe we can just test this corner case with a comment that it is not fully correct, but it is hard to define an attribute docstring the right way.

using class as a context is not a solution as well, because this code would pass:

```python
class A:
    def some(self, x):
        x.y = 1
        """Docstring."""
```

btw, let's maybe create a rule for placing docstings? this would solve the issue by enforcing the correct places for docstings.

Comment (vyhuholl) [reply wemake_python_styleguide/logic/tree/strings.py]: Fixed in https://github.com/vyhuholl/wemake-python-styleguide/commit/019a1ea4b1db5e7cce3cafd0cd8e7d03ec880532

A rule for placing docstrings is a genuinely good idea and a much better fix than any of this — enforce placement, and the definition problem goes away. I'll open a new issue about this.

**Review by sobolevn — APPROVED** [review]
Now this seems correct, thank you :)

**Comment by sobolevn** [comment]
Please, don't forget about new rule request for `docstring` positions :)
